csrc/multidevice/communication.cpp

```diff
@@ -97,32 +99,57 @@ T getInitialValue(c10d::ReduceOp::RedOpType op) {
 }
 
 bool hasRoot(CommunicationType type) {
-  return type == CommunicationType::Gather ||
-      type == CommunicationType::Scatter || type == CommunicationType::Reduce ||
-      type == CommunicationType::Broadcast ||
-      type == CommunicationType::SendRecv;
+  switch (type) {
+    case CommunicationType::Gather:
+    case CommunicationType::Scatter:
+    case CommunicationType::Reduce:
+    case CommunicationType::Broadcast:
+    case CommunicationType::SendRecv:
+      return true;
+    case CommunicationType::Allgather:
+    case CommunicationType::Allreduce:
+    case CommunicationType::ReduceScatter:
+      return false;
+    default:
+      NVF_ERROR(false, "unrecognized CommunicationType: ", type);
+  }
 }
 
 bool isReduction(CommunicationType type) {
-  return type == CommunicationType::Reduce ||
-      type == CommunicationType::Allreduce ||
-      type == CommunicationType::ReduceScatter;
+  switch (type) {
+    case CommunicationType::Reduce:
+    case CommunicationType::Allreduce:
+    case CommunicationType::ReduceScatter:
+      return true;
+    case CommunicationType::Gather:
+    case CommunicationType::Allgather:
+    case CommunicationType::Scatter:
+    case CommunicationType::Broadcast:
+    case CommunicationType::SendRecv:
+      return false;
+    default:
+      NVF_ERROR(false, "unrecognized CommunicationType: ", type);
+  }
 }
 
 } // namespace
 
 Communication::Communication(
     IrBuilderPasskey passkey,
     CommunicationType type,
-    DeviceMesh mesh,
+    TensorView* out,
+    TensorView* in,
     Team team,
     DeviceIdxType root,
     RedOpType red_op,
-    int64_t scattered_axis,
-    TensorView* input_tv,
-    TensorView* output_tv)
+    int64_t scattered_axis)
     : Expr(passkey) {
-  NVF_ERROR(mesh.size() > 0, "The mesh size must be greater than 0.");
+  NVF_ERROR(
+      in->getDeviceMesh().size() > 0,
+      "The input mesh size must be greater than 0.");
+  NVF_ERROR(
+      out->getDeviceMesh().size() > 0,
+      "The output mesh size must be greater than 0.");
```

Comment:
> We could allow one of the meshes to be empty and inferred from the other, but it feels less explicit and adds complexity to properly infer meshes.

There might be a few edge cases where this could be nice... For example, if a scheduler adds new exprs, it should re-run the sharding propagation pass, but it's easy to forget and we haven't updated all these passes everywhere. I would leave this as an error for now, but something to consider.